### ISSUE TITLE
Scalable solution for adding scripts to install things into a dev container

### ISSUE THREAD

**Issue by diberry** [microsoft/vscode-dev-containers#615]
- VSCode Version: 1.50.1 (system setup) d2e414d9e
- Local OS Version: Windows_NTx64 10.0.19042
- Name of Dev Container Definition with Issue: Node.js & TypeScript

Steps to Reproduce:

1. Add [Az cli install script](https://github.com/microsoft/vscode-dev-containers/blob/master/script-library/docs/azcli.md) as documented
2. Rebuild container
3. Load container
4. At VS Code terminal inside container, try command `az` to see if az is in path. `command not found`

I assume the containers are doing something that the provided script isn't. My goal is to get Az CLI on my mac in the easiest way possible and have it available while developing to the Azure cloud. Adding it to the container I was already using, based on the doc and script, seemed like the easiest way but isn't working. 

If I start up the Az CLI container, Az works on the terminal. 

Is there a reason that Az CLI isn't just added to all containers, which is what would be the most helpful.

**Comment by Chuxel** [comment]
That's odd given that all the CLI container does is run the script.  Can you point me to your Dockerfile?

> Is there a reason that Az CLI isn't just added to all container

Yes, the Azure CLI is actually massive - about 300mb. In addition, both GitHub Codespaces and Remote - Containers are intended to be cloud agnostic.

**Comment by diberry** [comment]
@Chuxel It is working now - I must have not rebuilt the container. Sorry about that. Is there anyway the Azure CLI install could be a question in the Dev Container selection process? So customers that need Azure CLI just have to answer a question and not go into the Dockerfile?

**Comment by Chuxel** [comment]
@diberry Realistically that opens up a big door for "pile on" that could slow startup time due to Docker build speed or ballooned image downloads (a critical focus at the moment) - we'd need to be okay with asking about AWS and Google CLIs as well which opens the door to a potential pile on of many different things.  These are intended to be the "minimal" config. 

However, what we've discussed the idea of snippets based on these scripts to make it easier for those that want to add different things in.  There's a number of others like SSH or Desktop support that are also of interest.

**Comment by Chuxel** [comment]
@bamurtaugh @chrmarti FYI - We may want to think about snippet support sooner than later. Ideally this would also be something that could be reused in Codespaces.

**Comment by bamurtaugh** [comment]
I think it'd make sense to start thinking about snippet support sooner as well, especially if it could be reused in Codespaces. 

Were we envisioning enabling automatically inserting each script via snippets when a user is in the Dockerfile? i.e. a user could type `azcli` in the Dockerfile (or whichever name we choose for the snippet) and it would automatically populate the file with the script to install the Azure CLI? 

In the [current guidelines](https://github.com/microsoft/vscode-dev-containers/blob/master/script-library/docs/azcli.md), users need to add azcli-debian.sh to `.devcontainer/library-scripts` and then a couple lines of code to their Dockerfile. Would the snippet process just automate adding those lines of code to their Dockerfile, or also help with inserting the file/code for the script in some way? The script files can be rather long (i.e. [common-alpine](https://github.com/microsoft/vscode-dev-containers/blob/master/script-library/common-alpine.sh)), so it seemed like it might not make sense to insert all of that code directly in a Dockerfile (hence why they're scripts in the first place I suppose?).

**Comment by Chuxel** [comment]
> Would the snippet process just automate adding those lines of code to their Dockerfile, or also help with inserting the file/code for the script in some way? 

@bamurtaugh Yeah, definitely wouldn't want to drop the content in.  It would either add a file reference or we can also reference it via curl if that ends up being too much. e.g.

```
RUN apt-get update && export DEBIAN_FRONTEND=noninteractive && apt-get -y install --no-install-recommends curl ca-certificates \
    && bash -c "$(curl -fsSL "https://raw.githubusercontent.com/microsoft/vscode-dev-containers/v0.145.1/script-library/azcli-debian.sh")" 
```

It just becomes a bit more complex a snippet and we have to make sure the version number (v0.145.1) in the URL is correct when added. We'd also need to be aware that this model assumes GH is up and responsive while adding the file does not.  That said, the file model may need a command instead of a snippet - I haven't looked yet.

**Comment by diberry** [comment]
As a fix for the short term, could the script be on the container and in the docker file but just commented out with explanatory text so someone doesn't have to go looking for it ?

**Comment by chrmarti** [comment]
Snippets might be hard to do well because they are either too low-level for a new user to correctly combine or too high-level to blend into an existing Dockerfile.

**Comment by diberry** [comment]
If there a way to have the RUN command in the .devcontainer file, commented out like forwarded ports, but there in case someone needs it?

**Comment by Chuxel** [comment]
@diberry The problem is that there are a lot of these cases, so we can't just do one, so we need to think about a scalable solution to the broader problem. There's over 20 scripts already.

@chrmarti The challenge here is that I think we do need to come up with a solution. Largely right now, we'll tell people to put scripts in a folder and then copy-paste a Dockerfile snippet. Most won't have any understanding of what it does particularly if they're new to Linux and bash scripting which is sort of its own little universe.

I could see an option out of the add devcontainer config flow to ask you whether you want to add other scripts, and then allow you to pick from a list that just drops command chunks in.  That might make it a bit more of a controllable set of permutations.

**Comment by bamurtaugh** [comment]
Similar to how we have the check box to install Node.js, it feels like letting users check off if they'd like to install something like the Azure CLI and then adding that script for them automatically could be a nice flow.

![image](https://user-images.githubusercontent.com/25310137/97373999-7f8aae00-1874-11eb-9a3d-10ac6dc929b5.png)

If feasible, this could help alleviate the concern about new users not knowing how to add the snippets? Although we'd still want to work on the formatting to address the concern about blending into an existing Dockerfile.

**Comment by Chuxel** [comment]
@bamurtaugh Yeah - that's kind of what I was getting at. What we have now is "common" options - there's then "additional" options that are not in the core definition and may make the build time long even if we've got an optimized image. We want to signal that this will happen, so we'd need to figure out how to make that clear.

FWIW - largely all the build args do is run the same snippets with just an "IF" statement around the script run.

**Comment by chrmarti** [comment]
We could add a marker `# [Options]` in the Dockerfiles and offer the user the library scripts to pick from. We could then copy the picked scripts to `.devcontainer/library-scripts` and add a `RUN` statement after the marker running those scripts.
- The marker would allow for positioning at the right place in the Dockerfile.
- We could list the compatible scripts in case we don't want all scripts to be offered for all definitions.
- If we put all in a single `RUN` statement, we could have 'initialize' and 'cleanup' scripts wrapping the other library scripts for optimizations.

**Comment by bamurtaugh** [comment]
I think the flow you mentioned makes sense and fits into the Dockerfile well. It looks like support was added for this flow already, which is really useful!

I had a couple of questions as I used it (but I'm thinking it's likely still under development so these are getting worked out).

**Flow 1: I added a C# dev container and didn't check off the Azure CLI.**

![image](https://user-images.githubusercontent.com/25310137/97476325-8284ad80-190b-11eb-8c66-57bc8be6bf04.png)

I didn't anticipate the library-scripts folder to be created or the Azure CLI library script to be copied over since I didn't select it:

![image](https://user-images.githubusercontent.com/25310137/97476429-9d572200-190b-11eb-8628-ed8f8f014692.png)

As I would've expected, the Dockerfile sets the Azure CLI's arg to be false. One thought is Dockerfiles could get long with quite a few scripts if we have to even list the false ones, but I'm thinking that's where your statement about compatible scripts comes into play.

![image](https://user-images.githubusercontent.com/25310137/97476467-a942e400-190b-11eb-8d49-36e738b0d656.png)

**Flow 2: The next time I checked off installing the Azure CLI.**

![image](https://user-images.githubusercontent.com/25310137/97476674-ef984300-190b-11eb-8d69-fc873a56451a.png)

A library-scripts folder was again successfully created with the Azure CLI library script. However, the Dockerfile still appears to have the ARG set as false:

![image](https://user-images.githubusercontent.com/25310137/97476770-0c347b00-190c-11eb-9052-99c876df1d49.png)

So when I reopen in container, `az` isn't recognized:

![image](https://user-images.githubusercontent.com/25310137/97477547-fffced80-190c-11eb-8f61-3fab65ec9f8b.png)

**Flow 3: Add a Python dev container instead of C#.**

I noticed the Azure CLI wasn't an option - is this also an instance of filtering for just compatible scripts, so it was deemed the Azure CLI script wasn't compatible for this definition? 

![image](https://user-images.githubusercontent.com/25310137/97477783-49e5d380-190d-11eb-90ca-b013c2d4f864.png)

**Comment by Chuxel** [comment]
@bamurtaugh To recap:

1. **Dev containers include common options** - .NET is the only non-Azure definition that includes it. For .NET has a strong tie to Azure and is a MS platform. Python and Node have healthy representation in all clouds and are not. If this is confusing, we should remove the Azure CLI from the .NET definition too.
2. **For common options, even if you made this choice now you can change your chose later** via the args in devcontainer.json. The files are there to support that scenario.
3. There are a ton of options we **could** include - that list is massive. This is the part that we'd need a different solution or you'd end up with a massive Dockerfile with 23 options. So, instead, we'd want a "Start with X, then add Y".

**Comment by chrmarti** [comment]
@bamurtaugh Just to clarify: My idea in https://github.com/microsoft/vscode-dev-containers/issues/615#issuecomment-717771518 isn't implemented yet. We currently use the approach with the build args which have default values in the Dockerfile that can be overridden from the devcontainer.json (so your choice ends up in that). `az` should work with what you did, have you tried rebuilding the container?

@Chuxel To clarify further: We would have only one `# [Options]` (plural, maybe there are better names) marker per Dockerfile and we would need a way to extract a label from each script for the UI. We could mark some scripts as popular and then offer the others only on request to make this scale better (in the future we could also try to rank them based on folder content).

That section in the Dockerfile could end up looking like:
```Dockerfile
# [Options]
COPY library-scripts /tmp/
RUN cd /tmp/library-scripts && \
    ./initialize-debian.sh && \
    ./azcli-debian.sh && \
    ./ghcli-debian.sh && \
    ./cleanup-debian.sh
```

If the user picked none of the library script options it would reduce to the marker (or nothing unless we want to be able to rerun that config step on the existing files).

**Comment by Chuxel** [comment]
@chrmarti Scripts need `ENV` vars and in some cases `ENTRYPOINT` changes.

I think a good one to focus on for the conversation is desktop support.

```Dockerfile
COPY library-scripts/desktop-lite-debian.sh /tmp/library-scripts/
RUN apt-get update && bash /tmp/library-scripts/desktop-lite-debian.sh
ENV DBUS_SESSION_BUS_ADDRESS="autolaunch:" DISPLAY=":1" LANG="en_US.UTF-8" LANGUAGE="en_US.UTF-8"
ENTRYPOINT ["/usr/local/share/desktop-init.sh"]
CMD ["sleep", "infinity"]
```

I haven't found a way to eliminate the ENVs with a script since all of this needs to be globally available - maybe you have an idea.

The entrypoint is needed to actually start the desktop.  The CMD is required because "overrideCommand" disables both the entrypoint and the CMD, so you have to set that to false and then add sleep infinity.

SSH and others are like this too.  That's why I think this is snippets.

**Comment by chrmarti** [comment]
We could add the environment variables where they are added in a full-fledged Linux install: `~/.profile` and `~/.bashrc`. But then we need to source these and pay some startup cost for the shell. One alternative could be to have the scripts append to `/etc/devcontainer/environment` `KEY=value` pairs of environment variables and the extension would then pick these up. The downside of that would be that a manual `docker exec <container id>` by the user wouldn't have these variables (so we could additionally source that file in `~/.profile`).

To support multiple entrypoints we could look into using a simple init system. (Makes me wonder if growing towards full-fledged Linux installs is the way to go.)

**Comment by Chuxel** [comment]
@chrmarti Re init, "tiny" built into docker handles most of init's responsibilities (--init in runArgs), but we could adopt `supervisord` for the process management. It doesn't need to be PID 1 so the container doesn't have to run privileged or anything like that. You'd still need an ENTRYPOINT, but it would just be one. The thing is, that makes me question the behavior off `"overrideCommand":false`. I think in these cases, you want to leave the ENTRYPOINT and just override CMD, but we could just set that to false and add a CMD if these scripts were used.

I avoided it to try to make sure image sizes stayed small since that's a major factor in initial create times as we encourage more publishing of images. We'd need to see how big it actually gets - it does require python, but so does git and many apt tools, so it may not be too big in practice.

For env, most scripts add content to `/etc/bash.bashrc` since this doesn't get in the way of any user config or dotfiles that update `~/.bashrc` (we could use `/etc/profile.d` technically as well if we only cared about login shells).

The trick is that these env vars need to be globally available in the `ENTRYPOINT` (or in  `supervisord` which would be kicked off from an ENTRYPOINT) and related child processes.  They also need to be available for user processes after that. I haven't found a way to do that since all the "rc" options require a login or interactive shell and ENTRYPOINT fires before VS Code attaches.

**Comment by chrmarti** [comment]
We wouldn't need a fully featured init system and could just place scripts in a folder that the entrypoint then runs sequentially. (Or using number prefixes we could run those with the same number prefix in parallel.)

The entrypoint could launch a shell to get the environment variables. Since the entrypoint usually runs services and the environment variables in `.bash_profile` / `.bashrc` are more relevant to users, the required environment variables might be few. Could you give some examples?

I agree `"overrideCommand"` should be more specific. One non-breaking way to do this would be to introduce `"overrideEntrypoint"` and make it default to `"overrideCommand"`.

**Comment by Chuxel** [comment]
> We wouldn't need a fully featured init system and could just place scripts in a folder that the entrypoint then runs sequentially. (Or using number prefixes we could run those with the same number prefix in parallel.)

Yep. Agreed. I'd considered just doing that with the /etc/init.d folder (firing `<script> start`).

> Could you give some examples?

The list above is the core example. In order for most UI based applications to run properly, all four of these need to be present:
```
DBUS_SESSION_BUS_ADDRESS="autolaunch:" 
DISPLAY=":1"
LANG="en_US.UTF-8"
LANGUAGE="en_US.UTF-8"
```
The first ensures DBUS automatically is setup, the second establishes which display everything should use, and the other two  are required for GNOME based applications that support multiple languages like nautilus or gnome-terminal. If any of these are missing, things fail. Both the entrypoint and every user shell require them.

**Comment by chrmarti** [comment]
Ack, how about appending KEY=VALUE pairs into `/etc/devcontainer/environment` or similar and read that in the entrypoint and from the extension? That will be fast because it doesn't require a shell. A manual `docker exec` by the user will get some or all of them from the `~/.profile` and `~/.bashrc` (and we can continue thinking about making that more seamless if needed too).

**Comment by Chuxel** [comment]
@chrmarti Do you know if Codespaces implements that? We'd need to stick with things that can be done there too. Also, when is the file created? Would it be overwritten if it already existed?

**Comment by chrmarti** [comment]
That would be a new file and Codespaces would have to implement support for it too.

One alternative would be to have metadata on each script file that can include environment variables. These are then either copied to the Dockerfile's `Options` section or the devcontainer.json's `containerEnv` property if the script is included.

**Comment by chrmarti** [comment]
I agree that adding everything to the Dockerfile seems like a good option. The main downside being the potentially long list of `ENV` instructions. I wonder how much that would change with `"userEnvProbe": "loginInteractiveShell"` (that also happens to be the VS Code Server's default behavior, so Codespaces is using this currently). E.g., the NVM environment variables wouldn't be needed as `ENV` instructions. (Any environment variables required in the entrypoint could still be written to a KEY=VALUE file to avoid having to source the shell scripts there.)

I realize we can further simplify the generated part of the Dockerfile by using numeric prefixes on the script files (in case there are any sort of dependencies) and have a `run.sh` that simply runs them in lexicographical order:
```
# [Options]
COPY library-scripts /tmp/
RUN /tmp/library-scripts/run.sh
# [/Options]
```

It might also be easiest to hold metadata, like names, files and ENV variables, in a single JSON file in the library-scripts folder:
```JSONC
[
  {
    "name": "Azure CLI",
    "files": ["40_azcli-debian.sh"],
    "distro": "debian"
  },
  {
    "name": "GitHub CLI",
    "files": ["40_ghcli-debian.sh"],
    "distro": "debian"
  },
  ...
]
```

**Comment by Chuxel** [comment]
>  I wonder how much that would change with "userEnvProbe": "loginInteractiveShell" (that also happens to be the VS Code Server's default behavior, so Codespaces is using this currently). E.g., the NVM environment variables wouldn't be needed as ENV instructions. (Any environment variables required in the entrypoint could still be written to a KEY=VALUE file to avoid having to source the shell scripts there.)

Yeah, I have been thinking along the same lines - I know we didn't change this as the default, but I continue to wonder leaving it as it was is more trouble than it is worth. I actually didn't know that Codespaces was running in this mode by default, so for consistency, that alone makes it valuable.

> I realize we can further simplify the generated part of the Dockerfile by using numeric prefixes on the script files

Yeah, that's true given that is effectively what happens with `rc.d` scripts and we'd likely need to do that for the `ENTRYPOINT` script if we wanted to only have one there anyway. That said, we'd have to tweak the auto-copy in the build system to match filename excluding the prefix if we wanted to use this same mechanism for things that should get built into image. But, that's not a huge change.

> A single JSON file in the library-scripts folder

Do you mean in the `/script-library` folder? One thing that has been a bit of a problem in this repo is upkeep with common content which is what led to `common-debian.sh` got created many months ago. If we can keep it in one file (or a set of files under a `/script-library/metadata` or something), that will reduce risk of breakage. We also may want to avoid json if we want it to be parsed by a bash script so there's no hard dependency on jq (which is installed by common-debian.sh, but still).

**Comment by chrmarti** [comment]
> I actually didn't know that Codespaces was running in this mode by default, so for consistency, that alone makes it valuable.

`"loginInteractiveShell"` is the VS Code Server's default (Codespaces doesn't actively do anything). We turned that off to fix https://github.com/microsoft/vscode-remote-release/issues/544. We need to investigate a new fix if we change this back. We later introduced `"userEnvProbe"`.

> > A single JSON file in the library-scripts folder
> 
> Do you mean in the `/script-library` folder? One thing that has been a bit of a problem in this repo is upkeep with common content which is what led to `common-debian.sh` got created many months ago. If we can keep it in one file (or a set of files under a `/script-library/metadata` or something), that will reduce risk of breakage. We also may want to avoid json if we want it to be parsed by a bash script so there's no hard dependency on jq (which is installed by common-debian.sh, but still).

The JSON would help us drive the UI, I wouldn't use it in the shell scripts themselves. I think this can be a single JSON file for simplicity. I imagine each such feature presented to the user would have an entry in the JSON file and that entry would come with a name for the UI and a list of files that need to be copied from the script-library to the user's project if that feature is selected.

I realize that "script-library" now means "scripts for the base images" (like `common-debian.sh`) and scripts for the user-selectable features, we can think about splitting these into two folders. We would ship the extension with all the scripts for the user-selectable features included, so we can copy these to the user's project folder when selected.

**Comment by Chuxel** [comment]
> The JSON would help us drive the UI, I wouldn't use it in the shell scripts themselves

This makes sense. This would also give us the ability to specify devcontainer.json properties that should be auto-set (e.g. `overrideCommand:false` or `overrideEntrypoint:false` if we add it). Probably better than trying to store all of this in comments at the end of the day. This can just get injected into the same repo npm package as metadata.

> ...we can think about splitting these into two folders. 

Yeah my mental model was that the `/script-library` folder gets added to the npm package out of this repo (assuming we don't do everything dynamically) to drive ad-hoc use.  The `library-scripts` folder for a definition in this repo just holds those that are actually part of the definition (e.g. `/containers/azure-terraform/library-scripts` for an example with a number of things that land in the customer's folder).  So either a sub-folder (`library-scripts/user`) another folder in the same location could contain the dynamic ones.

However, something would need to indicate which scrips were already run out of the broader set. For pre-built images, part or all of the `library-scripts` folder is in `.gitignore`. We could have the npm packaging process augment whatever json file is in `/script-library` with this information just by scanning the repo to avoid ongoing maintenance problems.

**Comment by chrmarti** [comment]
A definition's metadata could list exclusions for the library-scripts offered to the user. (Or is that the maintenance problem you are trying to avoid?)

**Comment by Chuxel** [comment]
@chrmarti Yeah we could add it to definition-manifest.json. I've got the build system setup so that it only tries to build it if the `build` section is in the json file.  That said, what I was thinking is we could actually just generate it automatically. Really the presence of the files tells us everything we need to know.

**Comment by Chuxel** [comment]
@chrmarti One possibility here would be to actually evolve the concept a bit and instead of having one metadata file, we give each script a bit more "life" and think about them as having a set of files. One is a manifest, another is the "install" script, and in concept we could have a "detect" script as well. The detect script could be used to see if script applies to the codebase. Starts to look a bit like a buildpack to build dev container rather than an application.

**Comment by PavelSosin-320** [comment]
Docker image and Docker application are two very different things. It is almost impossible to create single image application because of complexity of managing **more than one entry point** per container and managing more than **one process tree** inside container. Separate **configuration of more than one component** per Image is impossible absolutly.
Today standalone container applications in production are used only in IoT and limited resource area like Development boards and Routers.
Every engine able to run OCI (Docker) images is installed with built-in orchestration engine, Swarm at least, application builder, [Docker app plugin](https://docs.docker.com/app/working-with-app/) at least, Application deploier Docker stack deploy, at least.
Creation of single container application contradicts 3 main Docker concepts:
1. Fine scalability
2. Rolling update 
3. Deployment consistency
There is no restriction which prevents developer to create more that one image per application and put every tool of toolbox into separate image, side-car and deploy all this multicomponent development stack to runtime engine with single command.

**Comment by Chuxel** [comment]
@PavelSosin-320 From a deployment perspective this is correct. However we are trying to create an environment you can do development of arbitrary app types without forcing a hard pivot towards only using docker to build your app. It's a different use case.  Constructing custom images from building blocks is effectively what buildpacks do, so it's entirely possible if you start from a known base... which is what we'd do here.

**Comment by PavelSosin-320** [comment]
@Chuxel The non-read-only container makes the normal Docker scalability impossible. Such read-write containers also can't be restarted neither automatically nor manualy. In other words all user's work will be lost in the case of any minor failure of the container or infrastructure unless use of Docker container commit and tag new Image repeatedly.
The are several solutions how to extend Docker images effectivly, like [Multistage build](https://docs.docker.com/develop/develop-images/multistage-build/).  Some Dev pipeline providers have their own proprietly solution.
But what is common - Entire application must be repetatly deployable to the testing and production infrastructure.
The problem of multiple docker image versions has been solved too: RedHat Docker and Ubuntu Docker (Kube project) are installed with pre-installed and pre-configured local registry. The "latest" version of Dev-container image will contain everything what user desires to add to the base container.
So, the scalability requirement is fulfilled already by Docker without making containers read-write

**Comment by PavelSosin-320** [comment]
In the frame of Docker architecture this brilliant idea can't be implemented at all; what is the container state after container is "created" and scripts are finished to run? "undefined","inproper" like in Podman?
Podman provides better granularity for container state because Pod manages the container's instantiation. container can be "configured" before ready to serve. But the entry-script is a part of image metadata and available to the Pod. The understanding that Containers are ephemeral and their "overlay" filesystem is different from regular persistent volume or bind FS will be obvious to any user after 1st usage of docker container inspect command. 
This difference can't be ignored.

**Comment by PavelSosin-320** [comment]
@Chuxel Manifest is perfect!!!
BaseImage:
ImageTag: 
Pre-build: boolean
COPY | ADD : bollean
# Local repository is already configured in daemon.json - no need in additional configuration
Reading script from the Bind / Docker volume during initialization because Volume mount synchronization with Container start & entry-script can be very hard. No choice - it has to be put into the image. Maybe, definition of scripts folders like in normal Linux : init.d, systrmd/ is an option - to minimize configuration.
Every start /restart/scale of container causes  **/run/ folder erase, like in Linux**, unless Docker container commit is used all changes in Container FS by scripts, not volume persistancy will be lost!
The entry-script is invoked every time when container started/restarted.
In other words: Docker's assumption that Container FS is inmutable.
Changes can be applyed only on persistant volume data upon Container event STARTED, same as scaled to 1  is received.

**Comment by PavelSosin-320** [comment]
@Chuxel Is it responsible to install something after the container has been created??? When docker image is extended in a regular way every Dockerfile command adds an additional layer to the container but all undernet layers are shared among other containers and memory footprint growths very slow. It is easily controlled using the Docker's command Docker system df. The OOM situation can be prevented using a simple check after every run.
Enlargement of the running container during its lifetime after its creation simply adds entire occupied by the extension memory to every container instance. It can't be controlled and will result in OOM situation sooner or later. The management of Dockers OOM situation requires bachelor level knowledge in Docker and fine-tuning of Docker daemon parameters. In the case of a remote Docker engine running on a dedicated machine, it will require huge additional memory.

**Comment by Chuxel** [comment]
@PavelSosin-320 This is getting pretty off topic. I appreciate the input, but the topic here is about helping developers create custom Dockerfiles that meet their needsd.

**Comment by PavelSosin-320** [comment]
So, the requirements is about dockerfile customization, not the running container customization. Everything has to be done during Docker build and Docker container create will create read-only container instance. The customization will help in this layered technick desribed in the docker "Multistage build" documentation. Am I right?
The runtime cistomization is possible but it is clear that all persistant objects have to be created under /run/ or in any other tempfs

**Comment by w0otness** [comment]
I have a similar issue, where I need a way to provide CA certs to the bootstrap docker image, .

There is an (hacky) option to modify the volumeBootstrap.Dockerfile, to:
```
FROM alpine:3.13.2

COPY certs /certs/
RUN for filename in /certs/*.pem; do cat $filename >> /etc/ssl/certs/ca-certificates.crt; done && \
apk add --no-cache \
nodejs \
git \
openssh-client \
docker-cli \
docker-compose && \
for filename in /certs/*.pem; do cat $filename >> /etc/ssl/certs/ca-certificates.crt; done
```
but the COPY fails if the certs don't exist in the build context, I solved by putting a README into the dir so it has something to copy even if certs aren't added. You also have to run the for loop before and after apk so git doesn't SSL fail, as apk installing resets the certs (arg!).

The problem with that approach is that the bootstrap Dockerfile lives in the extension directory and whenever a new version of the Remote-Containers extension is released the manual changes are reverted.

Is it in scope of the current PR to provide a mechanism for making CA certs available to Dev Environment bootstrap images?, by having an option in the Settings UI to select CA certs/path, which then attaches a volume containing the certs to the alpine image, then runs update-ca-certificates. Clean. Can now git clone repo in corporate world. If not let me know and I'll create a separate issue.

**Comment by chrmarti** [comment]
@w0otness This is tracked as https://github.com/microsoft/vscode-remote-release/issues/3713.

**Comment by Chuxel** [comment]
Closing in favor of https://github.com/microsoft/vscode-remote-release/issues/5442 and https://github.com/microsoft/vscode-dev-containers/issues/1044.
